### PR TITLE
Add File#flock_* advisory locking methods to File.

### PR DESCRIPTION
**Pull request by technorama**

### CODE REVIEW

File: src/file/flock.cr [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    flock_shared blocking
+    begin
+      yield
+    ensure
+      flock_unlock
+    end
+  end
+
+  # Place a shared advisory lock. More than one process may hold a shared lock for a given file at a given time.
+  # Errno::EWOULDBLOCK is raised if blocking = false and an existing exclusive lock is set.
```
Comment (jhass) [inline src/file/flock.cr]: I would prefer "if \*blocking\* is set to \`false\`", people already try to call with keyword args using `foo(bar=baz)`

**Comment by technorama** [comment]
I do that all the time.  Why is the syntax for calling a method different from definition?

`foo("bar", baz = true)` seems to make more sense.

**Comment by asterite** [comment]
What if we name these `lock` instead of `flock`?

**Comment by ysbaddaden** [comment]
I agree, `lock_exclusive`, `lock_shared` and `unlock` would be enough.

**Comment by technorama** [comment]
Would have done that if there was a single type of file locking but there are 2.  Flock and fcntl locking are different (depending on os) and have different behaviors on the local filesystem vs nfs mounts.  For compatibility with other programs both will be needed.  I want the names of these methods to stand out so you know which API is used.

**Comment by asterite** [comment]
@technorama Sounds good. Could you rebase and we'll merge this?

**Pull request merged**